src/domain/usecases/SendNotificationDataValuesUseCase.ts

```diff
@@ -0,0 +1,347 @@
+    }
+
+    private async getOldDataValues(
+        dataSetId: string,
+        orgUnitId: string,
+        period: string,
+        dataValues: DataValue[]
+    ) {
+        const auditValues = await this.dataValuesRepository.getAudits({
+            dataSetIds: [dataSetId],
+            orgUnitIds: [orgUnitId],
+            periods: [period],
+        });
+
+        const dataValuesWithOldValue = dataValues.map(dv => {
+            const auditValuesByDe = auditValues.filter(
+                auditValue =>
+                    auditValue.dataElement.id === dv.dataElement &&
+                    auditValue.categoryOptionCombo.id === dv.categoryOptionCombo
+            );
+
+            const latestAuditValue = _(auditValuesByDe).first();
+            return {
+                ...dv,
+                oldValue: latestAuditValue?.value || "'  '",
+            };
+        });
+        return dataValuesWithOldValue;
+    }
+
+    private getDataSetConfig(dataSets: DataSet[], dataSetId: string, settings: Settings) {
+        const dataSet = dataSets.find(dataSet => dataSet.id === dataSetId);
+        if (!dataSet) {
+            throw Error(`Cannot find dataset: ${dataSetId}`);
+        }
+        return (
+            settings.dataSets[dataSet.id] ??
+            settings.dataSets[dataSet.code] ??
+            settings.dataSets[dataSet.name]
+        );
+    }
+
+    private async saveExecutionsWithEmail(
+        execution: NotificationDetail,
+        dataSetExecutions: NotificationStore,
+        key: string,
+        options: OptionsUseCase
+    ) {
+        const dateEmailSent = new Date().toISOString();
+        execution.lastDateSentEmail = dateEmailSent;
+        execution.lastUpdated = dateEmailSent;
+        log.debug(`Email sent ${dateEmailSent}`);
+
+        await this.saveExecutions(
+            {
+                ...dataSetExecutions,
+                [key]: {
+                    ...execution,
+                },
+            },
+            options
+        );
+    }
+
+    private async sendEmailToUsersInGroup(users: User[], subject: string, body: string) {
+        await promiseMap(users, async user => {
+            await this.notificationsRepository.send({
+                attachments: [],
+                subject,
+                body: {
+                    contents: body,
+                    type: "html",
+                },
+                recipients: [user.email],
+            });
+        });
+    }
+
+    private async getEmailTemplate(path: Path): Async<EmailTemplate> {
+        return jsonfile.readFile(path);
+    }
+
+    private async getUsersToNotify(dataValues: DataValue[], userGroup: string) {
+        const editorUsers = _(dataValues)
+            .map(dv => dv.storedBy)
+            .uniq()
+            .value();
+
+        const usersInGroup = await this.userRepository.getFromGroup([userGroup]);
+        return usersInGroup.filter(user => !editorUsers.includes(user.username));
```

Comment:
Not a problem here, just a general note for bigger data structure: filter+includes is a nested O(n), O(m) loop, so O(n*m). Solution: Convert the first structure to O(1), for example, editorUsers can be a `Set`, the the includes (has) it's O(1) and the full loop is O(n)